### ISSUE TITLE
Rule Proposal: Enforce a maximum number of statements per line

### ISSUE THREAD

**Issue by scriptdaemon** [eslint/eslint#5424]
`1. When the rules will warn. Include a description as well as sample code.`

`max-statements-per-line`: There should be a rule to warn the developer when any line(s) of their code contains more than a maximum allowed number of statements (ignoring variable declarations in for type specifiers).

Passing code:
```js
    var test, test2, test3;

    for (var i = 0; i < someNum; ++i) {
        // Code
    }

    function doSomething() {
        var obj = {
            prop1: 'value',
            prop2: 'otherValue'
        };

        console.log(obj.prop1);
    }
```
Not passing code:
```js
    var test; var test2; var test3;

    function doAThing() {
        doAnotherThing(); return true;
    }

    switch(variable) {
        case 'value1':
            doSomething(); break;
        default:
            doSomethingElse(); break;
    }
```

`2. Whether the rule prevents an error or is stylistic.` Stylistic.

`3. Why the rule should be in the core instead of creating a custom rule.`

I think this will enforce better readability and make it easier to visualize changes, plus, from what I've seen, it seems like a common best practice principle (in general, not just JS).

`4. Are you willing to create the rule yourself?` Yes.

**Comment by eslintbot** [comment]
Hi @scriptdaemon, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by platinumazure** [comment]
Hi @scriptdaemon, you've answered some of the questions we usually need answered, but there are a couple more. Please take a look at http://eslint.org/docs/developer-guide/contributing/new-rules#proposing-a-rule for more info. Thanks!

**Comment by scriptdaemon** [comment]
I updated the request to include the question I didn't answer and added the questions into the post to specify where I answered each question.

**Comment by doberkofler** [comment]
This is a rule I would also look forward to having in the core set of rules.

**Comment by mysticatea** [comment]
Sounds good to me.

**Comment by ilyavolodin** [comment]
I think that's generic enough of a request to fit into the core.

**Comment by scriptdaemon** [comment]
While brainstorming how to go about implementing this rule, I've come across several cases where I'd like some feedback.

Should this rule warn for the following code: `if (condition) statement;`?

If yes to the first question, what about: `statement; if (condition) ...`?

If yes to the first question, should we consider `else`, `catch`, and / or `while` (in a `DoWhileStatement`) as their own "statements"?

Should this rule warn when a statement follows on the same line as a closing brace?

**Comment by mysticatea** [comment]
In my opinion, this rule should warn statements when we cannot remove semicolon(s).

**Comment by BYK** [comment]
:+1: to @mysticatea - I think that's a very neat and good way to describe it.

**Comment by michaelficarra** [comment]
@mysticatea No, that's not the right way to look at it.

```js
if(a){} if(b){}
```

Two statements, one line, no semicolons.

**Comment by nzakas** [comment]
I'd alter this proposal a bit and call it `max-statements-per-line` and let the max be configurable. Hardcoding a max will undoubtedly give us trouble in the future.

**Comment by btmills** [comment]
My first instinct would be to implement this on blocks, iterating children and verifying that one child's ending line is less than the start line of the next child. This would catch `if (a) { } if (b) { }` but allow `if (a) return b;`.

**Comment by michaelficarra** [comment]
@btmills That sounds like the right approach to me. It should apply for all `[ Statement ]` in this document: https://github.com/estree/estree/blob/master/spec.md.

**Comment by mysticatea** [comment]
Oh @michaelficarra is right.
Never mind my comment.

**Comment by scriptdaemon** [comment]
I like @nzakas proposal, and it would be trivial to adjust based on my initial thoughts regarding how to do this.

Essentially, my idea was to create an array where the index is a corresponding line in a file, and the value would be the number of statements on that line / index. After that's done, iterate through the array and warn if the value is greater than the max allowed.

**Comment by michaelficarra** [comment]
@scriptdaemon No, you should follow the advice of @btmills.

**Comment by scriptdaemon** [comment]
@michaelficarra As per @btmills method, what would be the easiest way to allow for a max number of statements rather than just one then?

**Comment by michaelficarra** [comment]
@scriptdaemon Something like this:

```js
function enforceMaxStatements(listofStatements) {
  var numberOfStatementsOnThisLine = 0;
  if (listOfStatements.length < 2) return;
  var lastStatementLine = listOfStatements[0].loc.end.line;
  for (var i = 1, l = listOfStatements.length; i < l; ++i) {
    var currentStatement = listOfStatements[i];
    if (currentStatement.loc.start.line <= lastStatementLine) {
      ++numberOfStatementsOnThisLine;
    } else {
      numberOfStatementsOnThisLine = 1;
      lastStatementLine = currentStatement.loc.end.line;
    }
    if (numberOfStatementsOnThisLine > maxStatementsPerLine) {
      report(currentStatement);
    }
  }
}
```

**Comment by nzakas** [comment]
@scriptdaemon we don't rename rules, as that's an unnecessary breaking change.
